quinn-proto/src/connection/mtud.rs

```diff
@@ -350,6 +335,80 @@ impl SearchState {
     }
 }
 
+#[derive(Clone)]
+struct BlackHoleDetector {
+    /// Counts suspicious packet loss bursts since the last time a packet with size equal to the
+    /// current MTU was acknowledged
+    ///
+    /// Suspicious packet loss bursts are those that start with a packet of size >
+    /// BASE_UDP_PAYLOAD_SIZE
+    loss_burst_counter: u8,
+    /// The largest acked packet of size `current_mtu`
+    largest_acked_mtu_sized_packet: Option<u64>,
+}
+
+impl BlackHoleDetector {
+    fn new() -> Self {
+        Self {
+            loss_burst_counter: 0,
+            largest_acked_mtu_sized_packet: None,
+        }
+    }
+
+    pub(crate) fn on_probe_acked(&mut self) {
+        // We know for sure the path supports the current MTU
+        self.loss_burst_counter = 0;
+    }
+
+    pub(crate) fn on_non_probe_acked(
+        &mut self,
+        current_mtu: u16,
+        packet_number: u64,
+        packet_bytes: u16,
+    ) {
+        // Reset the black hole counter if a packet the size of the current MTU or larger
+        // has been acknowledged
+        if packet_bytes >= current_mtu
+            && self
+                .largest_acked_mtu_sized_packet
+                .map_or(true, |pn| packet_number > pn)
+        {
+            self.loss_burst_counter = 0;
+            self.largest_acked_mtu_sized_packet = Some(packet_number);
+        }
+    }
+
+    /// Notifies the [`BlackHoleDetector`] of a packet loss burst
+    pub(crate) fn on_packet_loss_burst(
+        &mut self,
+        leading_packet_number: u64,
+        leading_packet_size: u16,
```

Comment:
I assume the reason for looking only at the leading packet size is that it keeps the code in s2n-quic simple and provides enough robustness. However, in our case, I think there is no additional complexity in making it more consistent, by ensuring only MTU-sized packets were lost instead of only looking at the leading one. I'll implement it 😉